src/cc/api/BPF.cc

```diff
@@ -164,22 +168,37 @@ StatusTuple BPF::detach_all() {
     return StatusTuple(0);
 }
 
-StatusTuple BPF::attach_kprobe(const std::string& kernel_func,
-                               const std::string& probe_func,
-                               uint64_t kernel_func_offset,
-                               bpf_probe_attach_type attach_type,
-                               int maxactive) {
+StatusTuple BPF::attach_kprobe_(const std::string& kernel_func,
+                                const std::string& probe_func,
+                                const std::string& cgroup_name,
+                                uint64_t kernel_func_offset,
+                                bpf_probe_attach_type attach_type,
+                                int maxactive) {
   std::string probe_event = get_kprobe_event(kernel_func, attach_type);
   if (kprobes_.find(probe_event) != kprobes_.end())
     return StatusTuple(-1, "kprobe %s already attached", probe_event.c_str());
 
   int probe_fd;
   TRY2(load_func(probe_func, BPF_PROG_TYPE_KPROBE, probe_fd));
 
+  int cgroup_fd = -1;
+  if (!cgroup_name.empty()) {
+    // Compose a full path of cgroup
+    std::string cgroup_path = "/sys/fs/cgroup/perf_event/" + cgroup_name;
```

Comment:
Let's not get into speculating where is user's `cgroupfs` mounted. Just pass in the user-opened FD. If you open the FDs here we'll also need to properly close them, which is complicated.

Reply:
That works for me too if this is what you want. My concern is that it would be awkward to get a file descriptor for some languages, but certainly all languages should be able to handle string well ;-)

I know `cgroupfs` could be mounted anywhere before, but almost of distribution have it mounted to `/sys/fs/cgroup`. If you're concerning about this, we can read and parse `/proc/mounts` to get the exact location of cgroup mount, which is probably better than passing file descriptor.

Reply:
Considering sometimes cgroup could have some hierarchies like `/sys/fs/cgroup/system.slice/system-mysqld.slice/mysqld@3300.service`. You want to trace `mysqld@3300.service` which has a few processes in it. Just cgroup name `mysqld@3300.service` as it is possible to have the same name in other hierarchy.

Also, let us consider the python tools interface. The tools could accept a complete path like `/sys/fs/cgroup/system.slice/system-mysqld.slice/mysqld@3300.service` or a relative path `system.slice/system-mysqld.slice/mysqld@3300.service`. We should not bother tools writer with managing cgroup_fd. So bcc python library should open/close cgroup_fd as appropriate. 

So in my opinion, the C/C++ API should be cgroup_fd based and python API, which should be cgroup_path based. python library should handle open/close cgroup_fd and then using C++ binding. This way, we keep C/C++ library a little bit concise.

Reply:
which C/C++ API is used by python? Is it `src/cc/api/BPF.[cc,h]` or `src/cc/libbpf.c`?

Reply:
See bcc/src/python/bcc/libbcc.py.

Reply:
Agreed with @yonghong-song . Let's keep the C / C++ API take cgroup FD and we can figure out Python interface later.